File: packages/schema/src/authoring.ts
import type { ChoiceOption, ExtractionIssue, FieldNode, FormDefinition, ValidationRule } from "./generated";
import type { RuntimeDocumentBehavior, RuntimeNodeBehavior } from "./runtime";

export type ProjectStatus = "draft" | "published";
export type DocumentClass = FormDefinition["documentClass"];
export type ReviewStatus = FormDefinition["reviewStatus"];
export type SemanticType = FieldNode["semanticType"];

export interface AuthoringField {
  id: string;
  dispatchKey?: string | null;
  stableKey: string;
  label: string;
  helpText?: string | null;
  semanticType: SemanticType;
  required: boolean;
  confidence?: number | null;
  options: ChoiceOption[];
  validations: ValidationRule[];
  layoutHints: Record<string, string>;
  rendererHints: Record<string, string>;
  sourcePriority: string[];
  sourceConflicts: string[];
  lineage: string[];
  sourceFieldIds: string[];
  provenanceAnchorIds: string[];
  runtime?: RuntimeNodeBehavior | null;
}

export interface AuthoringGroup {
  id: string;
  dispatchKey?: string | null;
  label: string;
  description?: string | null;
  layoutHints: Record<string, string>;
  rendererHints: Record<string, string>;
  lineage: string[];
  sourceGroupIds: string[];
  provenanceAnchorIds: string[];
  fields: AuthoringField[];
  runtime?: RuntimeNodeBehavior | null;
}

export interface AuthoringSection {
  id: string;
  dispatchKey?: string | null;
  title: string;
  description?: string | null;
  layoutHints: Record<string, string>;
  lineage: string[];
  sourceSectionIds: string[];
  provenanceAnchorIds: string[];
  groups: AuthoringGroup[];
  fields: AuthoringField[];
  runtime?: RuntimeNodeBehavior | null;
}

export interface AuthoringStep {
  id: string;
  dispatchKey?: string | null;
  title: string;
  description?: string | null;
  kind: string;
  layoutHints: Record<string, string>;
  sourcePageIds: string[];
  provenanceAnchorIds: string[];
  sections: AuthoringSection[];
  runtime?: RuntimeNodeBehavior | null;
}

export interface AuthoringDocument {
  id: string;
  dispatchKey?: string | null;
  title: string;
  documentClass: DocumentClass;
  reviewStatus: ReviewStatus;
  targetRuntime: string;
  visualBaseline: string;
  sourcePriority: string[];
  sourceConflicts: string[];
  steps: AuthoringStep[];
  metadata: Record<string, string>;
  runtime?: RuntimeDocumentBehavior | null;
}

export interface ProjectSourceContext {
  conversionId: string;
  filename: string;
  documentClass?: DocumentClass | null;
  reviewStatus: ReviewStatus;
  confidence: number;
  extractorPath: string[];
  notes: string[];
  issues: ExtractionIssue[];
  importedDraft: FormDefinition;
}

export interface AuthoringProjectRecord {
  id: string;
  name: string;
  status: ProjectStatus;
  targetRuntime: string;
  visualBaseline: string;
  sourceConversionId: string;
  documentClass?: DocumentClass | null;
  currentRevisionId?: string | null;
  revisionCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface AuthoringProjectDetail {
  project: AuthoringProjectRecord;
  document: AuthoringDocument;
  sourceContext: ProjectSourceContext;
}

export interface ProjectRevision {
  id: string;
  projectId: string;
  createdAt: string;
  note: string;
  document: AuthoringDocument;
}

// ---- Behavior Library (Phase 1A schema-only; Phase 1C wires the editor) ----

export type BehaviorLibraryParameterType =
  | "string"
  | "number"
  | "boolean"
  | "enum"
  | "nodeRef"
  | "eventRef"
  | "fieldKey"
  | "list";

export interface BehaviorLibraryParameterConstraints {
  regex?: string;
  min?: number;
  max?: number;
}

export interface BehaviorLibraryParameter {
  key: string;
  type: BehaviorLibraryParameterType;
  label: string;
  description?: string;
  required?: boolean;
  default?: unknown;
  options?: string[]; // for enum
  itemType?: BehaviorLibraryParameterType; // for list<T>
  constraints?: BehaviorLibraryParameterConstraints;
}

export type BehaviorLibraryScope = "system" | "project";

export type BehaviorLibraryCategory =
  | "validation"
  | "visibility"
  | "host"
  | "events"
  | "data"
  | "repeatables"
  | "custom";

export interface BehaviorLibraryEntryI18nText {
  name: string;
  description: string;
}

export interface BehaviorLibraryEntry {
  id: string;
  name: string;
  description: string;
  category: BehaviorLibraryCategory;
  scope: BehaviorLibraryScope;
  revision: number;
  icon?: string;
  parameters: BehaviorLibraryParameter[];
  // Restricts applicability. Empty array = any node type.
  bindsTo: string[];
  // Listener template with {{paramKey}} placeholders. Refs in template MUST
  // reference parameter tokens, not literal ids. Phase 1A leaves this opaque
  // (Phase 1C wires the editor).
  template: unknown;
  i18n?: { en: BehaviorLibraryEntryI18nText; [lang: string]: BehaviorLibraryEntryI18nText };
}

File: apps/api/src/form_builder_api/models/authoring.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Literal
from uuid import uuid4

from pydantic import Field

from .base import CamelModel
from .canonical import (
    ChoiceOption,
    DocumentClass,
    ExtractionIssue,
    FormDefinition,
    ReviewStatus,
    SemanticType,
    ValidationRule,
)
from .runtime import RuntimeDocumentBehavior, RuntimeNodeBehavior


class ProjectStatus(str, Enum):
    DRAFT = "draft"
    PUBLISHED = "published"


class AuthoringField(CamelModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    dispatch_key: str | None = None
    stable_key: str
    label: str
    help_text: str | None = None
    semantic_type: SemanticType
    required: bool = False
    confidence: float | None = Field(default=None, ge=0, le=1)
    options: list[ChoiceOption] = Field(default_factory=list)
    validations: list[ValidationRule] = Field(default_factory=list)
    layout_hints: dict[str, str] = Field(default_factory=dict)
    renderer_hints: dict[str, str] = Field(default_factory=dict)
    source_priority: list[str] = Field(default_factory=list)
    source_conflicts: list[str] = Field(default_factory=list)
    lineage: list[str] = Field(default_factory=list)
    source_field_ids: list[str] = Field(default_factory=list)
    provenance_anchor_ids: list[str] = Field(default_factory=list)
    runtime: RuntimeNodeBehavior | None = None


class AuthoringGroup(CamelModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    dispatch_key: str | None = None
    label: str
    description: str | None = None
    layout_hints: dict[str, str] = Field(default_factory=dict)
    renderer_hints: dict[str, str] = Field(default_factory=dict)
    lineage: list[str] = Field(default_factory=list)
    source_group_ids: list[str] = Field(default_factory=list)
    provenance_anchor_ids: list[str] = Field(default_factory=list)
    fields: list[AuthoringField] = Field(default_factory=list)
    runtime: RuntimeNodeBehavior | None = None


class AuthoringSection(CamelModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    dispatch_key: str | None = None
    title: str
    description: str | None = None
    layout_hints: dict[str, str] = Field(default_factory=dict)
    lineage: list[str] = Field(default_factory=list)
    source_section_ids: list[str] = Field(default_factory=list)
    provenance_anchor_ids: list[str] = Field(default_factory=list)
    groups: list[AuthoringGroup] = Field(default_factory=list)
    fields: list[AuthoringField] = Field(default_factory=list)
    runtime: RuntimeNodeBehavior | None = None


class AuthoringStep(CamelModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    dispatch_key: str | None = None
    title: str
    description: str | None = None
    kind: str = "collect"
    layout_hints: dict[str, str] = Field(default_factory=dict)
    source_page_ids: list[str] = Field(default_factory=list)
    provenance_anchor_ids: list[str] = Field(default_factory=list)
    sections: list[AuthoringSection] = Field(default_factory=list)
    runtime: RuntimeNodeBehavior | None = None


class AuthoringDocument(CamelModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    dispatch_key: str | None = None
    title: str
    document_class: DocumentClass
    review_status: ReviewStatus = ReviewStatus.REVIEWED
    target_runtime: str = "va_web_form"
    visual_baseline: str = "va.gov"
    source_priority: list[str] = Field(default_factory=list)
    source_conflicts: list[str] = Field(default_factory=list)
    steps: list[AuthoringStep] = Field(default_factory=list)
    metadata: dict[str, str] = Field(default_factory=dict)
    runtime: RuntimeDocumentBehavior | None = None


class ProjectSourceContext(CamelModel):
    conversion_id: str
    filename: str
    document_class: DocumentClass | None = None
    review_status: ReviewStatus
    confidence: float = Field(ge=0, le=1, default=0)
    extractor_path: list[str] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)
    issues: list[ExtractionIssue] = Field(default_factory=list)
    imported_draft: FormDefinition


class AuthoringProjectRecord(CamelModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    name: str
    status: ProjectStatus = ProjectStatus.DRAFT
    target_runtime: str = "va_web_form"
    visual_baseline: str = "va.gov"
    source_conversion_id: str
    document_class: DocumentClass | None = None
    current_revision_id: str | None = None
    revision_count: int = Field(default=0, ge=0)
    created_at: datetime
    updated_at: datetime


class AuthoringProjectDetail(CamelModel):
    project: AuthoringProjectRecord
    document: AuthoringDocument
    source_context: ProjectSourceContext


class ProjectRevision(CamelModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    project_id: str
    created_at: datetime
    note: str
    document: AuthoringDocument


class ProjectPatch(CamelModel):
    name: str | None = None
    status: ProjectStatus | None = None


BehaviorLibraryParameterType = Literal[
    "string",
    "number",
    "boolean",
    "enum",
    "nodeRef",
    "eventRef",
    "fieldKey",
    "list",
]

BehaviorLibraryScope = Literal["system", "project"]

BehaviorLibraryCategory = Literal[
    "validation",
    "visibility",
    "host",
    "events",
    "data",
    "repeatables",
    "custom",
]


class BehaviorLibraryParameterConstraints(CamelModel):
    regex: str | None = None
    min: float | None = None
    max: float | None = None


class BehaviorLibraryParameter(CamelModel):
    key: str
    type: BehaviorLibraryParameterType
    label: str
    description: str | None = None
    required: bool = False
    default: object | None = None
    options: list[str] | None = None
    item_type: BehaviorLibraryParameterType | None = None
    constraints: BehaviorLibraryParameterConstraints | None = None


class BehaviorLibraryEntryI18nText(CamelModel):
    name: str
    description: str


class BehaviorLibraryEntry(CamelModel):
    id: str
    name: str
    description: str
    category: BehaviorLibraryCategory
    scope: BehaviorLibraryScope
    revision: int
    icon: str | None = None
    parameters: list[BehaviorLibraryParameter] = Field(default_factory=list)
    binds_to: list[str] = Field(default_factory=list)
    template: object  # opaque per Phase 1A
    i18n: dict[str, BehaviorLibraryEntryI18nText] | None = None
